components/feature/downloads/src/test/java/mozilla/components/feature/downloads/DownloadMiddlewareTest.kt

```diff
@@ -8,34 +8,181 @@ import android.app.DownloadManager.EXTRA_DOWNLOAD_ID
+        val applicationContext: Context = mock()
+        val downloadStorage: DownloadStorage = mock()
+        val downloadMiddleware = DownloadMiddleware(
+            applicationContext,
+            AbstractFetchDownloadService::class.java,
+            downloadStorage = downloadStorage,
+            coroutineContext = coroutineContext
+        )
+        val store = BrowserStore(
+            initialState = BrowserState(),
+            middleware = listOf(downloadMiddleware)
+        )
+
+        val download = DownloadState("https://mozilla.org/download", destinationDirectory = "")
+        store.dispatch(DownloadAction.AddDownloadAction(download)).joinBlocking()
+
+        store.dispatch(DownloadAction.RemoveDownloadAction(download.id)).joinBlocking()
+
+        verify(downloadStorage).remove(download)
+    }
+
+    @Test
+    fun `RemoveAllDownloadsAction MUST remove all downloads from the storage`() = runBlockingTest {
+        val applicationContext: Context = mock()
+        val downloadStorage: DownloadStorage = mock()
+        val downloadMiddleware = DownloadMiddleware(
+            applicationContext,
+            AbstractFetchDownloadService::class.java,
+            downloadStorage = downloadStorage,
+            coroutineContext = coroutineContext
+        )
+        val store = BrowserStore(
+            initialState = BrowserState(),
+            middleware = listOf(downloadMiddleware)
+        )
+
+        val download = DownloadState("https://mozilla.org/download", destinationDirectory = "")
+        store.dispatch(DownloadAction.AddDownloadAction(download)).joinBlocking()
+
+        store.dispatch(DownloadAction.RemoveAllDownloadsAction).joinBlocking()
+
+        verify(downloadStorage).removeAllDownloads()
+    }
+
+    @Test
+    fun `UpdateDownloadAction MUST update the storage when changes are meaningful`() = runBlockingTest {
+        val applicationContext: Context = mock()
+        val downloadStorage: DownloadStorage = mock()
+        val downloadMiddleware = DownloadMiddleware(
+            applicationContext,
+            AbstractFetchDownloadService::class.java,
+            downloadStorage = downloadStorage,
+            coroutineContext = coroutineContext
+        )
+        val store = BrowserStore(
+            initialState = BrowserState(),
+            middleware = listOf(downloadMiddleware)
+        )
+
+        val download = DownloadState("https://mozilla.org/download")
+        store.dispatch(DownloadAction.AddDownloadAction(download)).joinBlocking()
+
+        val downloadInTheStore = store.state.downloads.getValue(download.id)
+
+        assertEquals(download, downloadInTheStore)
+
+        var updatedDownload = download.copy(url = "updatedURL")
+        store.dispatch(DownloadAction.UpdateDownloadAction(updatedDownload)).joinBlocking()
+
+        verify(downloadStorage).update(updatedDownload)
+
+        updatedDownload = download.copy(url = "updatedURLAgain")
```

Comment:
nit: use a test case that can happen...this seems a bit contrived?

Reply:
updated!